lib/control-connection.js

```diff
@@ -773,6 +812,74 @@ ControlConnection.prototype.reset = function (callback) {
   });
 };
 
+/**
+ * Computes the highest supported protocol version collectively by the given hosts.
+ * 
+ * Considers the cassandra_version of the input hosts to determine what protocol versions
+ * are supported and uses the highest common protocol version amont them.
+ * 
+ * If hosts >= C* 3.0 are detected, any hosts older than C* 2.1 will not be considered
+ * as those cannot be connected to.  In general this will not be a problem as C* does
+ * not support clusters with nodes that have versions that are more than one major
+ * version away from each other.
+ *
+ * @param {Number} version Maximum version to consider.
+ * @param {Array.<Host>} hosts The hosts to determine highest protocol version from.
+ * @return {Number} Highest supported protocol version among hosts.
+ */
+ControlConnection.prototype._highestCommonProtocolVersion = function(version, hosts) {
+  let maxVersion = version;
+  // whether or not protocol v3 is required (nodes detected that don't support < 3).
+  let v3Requirement = false;
+  // track the common protocol version >= v3 in case we encounter older versions.
+  let maxVersionWith3OrMore = version;
+  const self = this;
+  hosts.forEach(h => {
+    if (!h.cassandraVersion || h.cassandraVersion.length === 0) {
+      self.log('warning', 'Encountered host ' + h.address + ' with no cassandra version,' +
+        ' skipping as part of protocol version evaluation');
+      return;
+    }
+
+    try {
+      const cassandraVersion = VersionNumber.parse(h.cassandraVersion);
+      self.log('verbose', 'Encountered host ' + h.address + ' with cassandra version ' + cassandraVersion);
+      if (cassandraVersion.compare(V3_0_0) >= 0) {
```

Comment:
Have you thought about how are we going to extend this behaviour on the DSE Driver?
Maybe we can move this logic to `protocol-version` module?

Reply:
Ah yes, that's a much better idea, will do that.  I had initially implemented it that way but reverted the logic back to control-connecton and not sure why.

In terms the of the DSE driver, I hope it should just be a matter of comparing the `dse_version` column and maybe falling back to `cassandra_version` if that isn't present (although I think it always is).